src/modules/daemonconfiguration/src/daemonconfiguration.rs

```diff
@@ -0,0 +1,380 @@
+#[repr(u8)]
+pub enum AutoStartStatus {
+    Other = 0,
+    Enabled = 1,
+    Disabled = 2,
+}
+
+// A daemon_config object with all possible setting types
+#[derive(Serialize, Deserialize, Debug, PartialEq, Eq)]
+#[serde(rename_all = "camelCase")]
+pub struct Daemon {
+    name: String,
+    state: State,
+    auto_start_status: AutoStartStatus,
+}
+
+pub trait SystemctlInfo {
+    fn get_daemons() -> Result<Vec<Daemon>, MmiError>;
+    fn list_unit_files() -> Result<String, MmiError>;
+    fn show_substate_property(name: &str) -> Result<String, MmiError>;
+    fn create_daemon(name: &str, auto_start_status: &str) -> Result<Daemon, MmiError>;
+}
+
+pub struct Systemctl {}
+
+impl SystemctlInfo for Systemctl {
+    fn list_unit_files() -> Result<String, MmiError> {
+        let output = {
+            Command::new("systemctl")
+                .args(["list-unit-files", "--type=service"])
+                .output()?
+        };
+        Ok(String::from(std::str::from_utf8(&output.stdout)?))
+    }
+
+    fn get_daemons() -> Result<Vec<Daemon>, MmiError> {
+        let systemctl_output = Self::list_unit_files()?;
+        lazy_static! {
+            static ref RE: Regex =
+                Regex::new(r"(?P<service>[\w_@\-\\\.]*)\.service\s+(?P<status>[\w-]+)").unwrap();
+        }
+        let mut services = Vec::new();
+        for service in RE.captures_iter(&systemctl_output) {
+            if service["service"].ends_with("@") {
+                continue;
+            }
+            let daemon = Self::create_daemon(&service["service"], &service["status"])?;
+            services.push(daemon);
+        }
+        Ok(services)
+    }
+
+    fn create_daemon(name: &str, auto_start_status: &str) -> Result<Daemon, MmiError> {
+        let substate = Self::show_substate_property(name)?;
+        if !substate.starts_with("SubState=") {
+            return Err(MmiError::SystemctlError);
+        }
+        let substate = &substate[9..];
+        let state = match substate {
+            "running" => State::Running,
+            "failed" => State::Failed,
+            "exited" => State::Exited,
+            "dead" => State::Dead,
+            _ => State::Other,
+        };
+        let auto_start_status = match auto_start_status {
+            "enabled" => AutoStartStatus::Enabled,
+            "disabled" => AutoStartStatus::Disabled,
+            _ => AutoStartStatus::Other,
+        };
+        Ok(Daemon {
+            name: String::from(name),
+            state: state,
+            auto_start_status: auto_start_status,
+        })
+    }
+
+    fn show_substate_property(name: &str) -> Result<String, MmiError> {
+        let output = {
+            Command::new("systemctl")
+                .args(["show", name, "--property=SubState"])
+                .output()?
+        };
+        Ok(String::from(std::str::from_utf8(&output.stdout)?))
+    }
+}
+
+#[derive(Default, Debug)]
+pub struct DaemonConfiguration {
```

Comment:
Consider moving the generic type to be at the struct level. This will ensure all instances of the struct will use the same underlying systemctl implementation (actual vs mock). instead of just get/set method.

Reply:
So did some experimentation and may have done it wrong, but then the struct needs to have fields of the generic type? (which would be unnecessary because an actual instance isn't needed, only the static methods). Also, leaving it get/set level prevents needing to specify the generic type for new() and getInfo()

Reply:
Instead of adding to the struct itself, it should be possible to add on the `impl` block for the struct: 

https://dev.to/somedood/generic-impl-blocks-are-kinda-like-macros-1aa0

*Also check out the optimization section of that article*

Reply:
Am I misunderstanding? impl<T> requires that the struct have a generic type (DaemonConfiguration<T>) so it'd have to be added to the struct. Also which optimization section are you referring to?

Reply:
Discussed offline, current implementation is fine practice (tie the generic to the method instead of the struct) for code cleanliness and abstraction